FullStorySegmentMiddleware/Classes/FSSuffixedProperties.m

```diff
@@ -0,0 +1,145 @@
+//
+//  FSSuffixedProperties.m
+//  Analytics
+//
+//  Created on 6/25/20.
+//  Copyright © 2020 FullStory. All rights reserved.
+//
+
+// FSSuffixedProperties comply with FS data requirements
+// Sample input Dictionary:
+// @{ @"item_name": @"Sword of Heracles",
+//    @"revenue": @2.95,
+//    @"features": @[@"Slay the Nemean Lion", @"Slay the nine-headed Lernaean Hydra."],
+//    @"products": @[
+//              @{@"product_id": @"productid_1",@"price": @1.11},
+//              @{@"product_id": @"productid_2",@"price": @2.22}
+//      ],
+//    @"mixed_array": @[@2,@4,@3.5,@"testmix",@[@"arr1", @"arr2"],@{@"objkey":@"objVal"},@4.5,@[@"arr3", @"arr4"]],
+//    @"int_array":@[@2,@4,@5,@10],
+//    @"bool_test": @true,
+//    @"nestedobjs_test":@{@"nestedkey":@"nestedval"},
+//    @"coupons_test":@[@{@"discount":@10.5},@{@"freeshipping":@true}]
+// }
+// Sample suffixedProperties:
+// @{ @"item_name_str": @"Sword of Heracles",
+//    @"revenue_real": @2.95,
+//    @"features_strs": @[@"Slay the Nemean Lion", @"Slay the nine-headed Lernaean Hydra."],
+//    @"products.product_id_strs": @[@"Sproductid_2", @"productid_1"],
+//    @"products.price_reals": @[@2.22, @1.11],
+//    @"mixedArr.objkey_str": @"objVal"
+//    @"mixedArr_ints":@[@2,@4]
+//    @"mixedArr_reals": @[@3.5, @4.5]
+//    @"mixedArr_strs": @[@"testmix", @"arr1", @"arr2", @"arr3", @"arr4"]
+//    @"int_array_ints":@[@2,@4,@5,@10],
+//    @"bool_test_int": @1,
+//    @"nestedobjs_test":@{@"nestedkey":@"nestedval"},
+//    @"coupons_test.price_real": @10.5,
+//    @"coupons_test.freeshipping_int": @1
+// }
+
+#import <Foundation/Foundation.h>
+#import "FSSuffixedProperties.h"
+
+@implementation FSSuffixedProperties
+
+- (id)init {
+    // init with empty properteis
+    return [self initWithProperties:nil];
+}
+
+- (id)initWithProperties:(NSDictionary *)properties {
+    // transform properties to comply with FS custom events requirement
+    // more info: https://help.fullstory.com/hc/en-us/articles/360020623234-FS-Recording-Client-API-Requirements
+    // TODO: Segment will crash and not allow props to have curcular dependency/nested or mixed arrays, but we should handle it here anyways
+
+    self.suffixedProperties = [[NSMutableDictionary alloc] init];
+
+    // Depth first search to iterate through potentially nested properties
+    NSMutableArray *stack = [[NSMutableArray alloc] initWithObjects:properties, nil];
+    while (stack.count > 0) {
```

Comment:
This is pretty minor, but you could use `[NSArray lastObject]` and do the fetch and while loop check do redoing this a bit:
```
NSDictionary *dict;
while ((dict = [stack lastObject]) != nil) {
```
Some people don't like the assignment/comparison at the same time like that. In any case I'd at least replace the next line `[stack objectAtIndex:(stack.count - 1)]` with `[stack lastObject]`, and replace `[stack removeObject:dict];` with `[stack removeLastObject];` since those are more readable, and removeLastObject doesn't need to scan the array.